Frameworks/CoreGraphics/CGContextCairo.mm

```diff
@@ -2011,7 +2011,32 @@
     // We need some special handling in transform as CoreText in iOS renders from bottom left but DWrite on Windows does top left.
     //     1. Scaling - to handle scaling once text has been scaled, apply translation to the scaled height
     //     2. Rotation - CoreText rotates text anti-clockwise but DWrite performs clockwise rotation
-    CGAffineTransform transform = CGAffineTransformConcat(curState->curTransform, curState->curTextMatrix);
+    CGAffineTransform transform = CGAffineTransformIdentity;
+
+    // Apply the text transformation (text position, text matrix) in text space rather than user space
+    // This means flipping the coordinate system,
+    // and applying the transformation about the center of the glyph run rather than about the baseline
+    // Flip and translate by the difference between the center and the baseline, apply text transforms, then flip and translate back
+    DWRITE_FONT_METRICS fontMetrics;
+    glyphRun->fontFace->GetMetrics(&fontMetrics);
+    float baselineCenterDelta = (fontMetrics.ascent - fontMetrics.descent) * glyphRun->fontEmSize / fontMetrics.designUnitsPerEm / 2;
+
+    // Transform to text space
+    // Technically there should be a horizontal translation to the center as well,
+    // but it's to the center of _each individual glyph_, as the reference platform applies the text matrix to each glyph individually
+    // Uncertain whether it's ever going to be worth it to support this using DWrite, so just ignore it for now
+    transform = CGAffineTransformConcat(CGAffineTransformMake(1, 0, 0, -1, 0, baselineCenterDelta), transform);
```

Comment:
Or rather: CGAffineTransform transform = CGAffineTransformMake(...);

---
In reply to: [84172077](https://github.com/Microsoft/WinObjC/pull/1188#discussion_r84172077) [](ancestors = 84172077)